common/src/main/java/com/viaversion/viabackwards/api/rewriters/LegacyBlockItemRewriter.java

```diff
@@ -43,70 +47,93 @@
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 public abstract class LegacyBlockItemRewriter<C extends ClientboundPacketType, S extends ServerboundPacketType,
     T extends BackwardsProtocol<C, ?, ?, S>> extends ItemRewriterBase<C, S, T> {
 
-    private static final Map<String, Int2ObjectMap<MappedLegacyBlockItem>> LEGACY_MAPPINGS = new HashMap<>();
-    protected final Int2ObjectMap<MappedLegacyBlockItem> replacementData; // Raw id -> mapped data
+    protected Int2ObjectMap<MappedLegacyBlockItem> replacementData = new Int2ObjectOpenHashMap<>(8); // Raw id -> mapped data
 
-    static {
-        JsonObject jsonObject = BackwardsMappingDataLoader.INSTANCE.loadFromDataDir("legacy-mappings.json");
-        for (Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
-            Int2ObjectMap<MappedLegacyBlockItem> mappings = new Int2ObjectOpenHashMap<>(8);
-            LEGACY_MAPPINGS.put(entry.getKey(), mappings);
-            for (Map.Entry<String, JsonElement> dataEntry : entry.getValue().getAsJsonObject().entrySet()) {
-                addMapping(dataEntry.getKey(), dataEntry.getValue().getAsJsonObject(), mappings);
+    protected LegacyBlockItemRewriter(T protocol, String mappedVersion) {
+        super(protocol, Type.ITEM1_8, Type.ITEM1_8_SHORT_ARRAY, false);
+        final JsonObject jsonObject = BackwardsMappingDataLoader.INSTANCE.loadFromDataDir("item-mappings-" + mappedVersion + ".json");
+        for (Map.Entry<String, JsonElement> dataEntry : jsonObject.entrySet()) {
+            final String key = dataEntry.getKey();
+            final JsonObject object = dataEntry.getValue().getAsJsonObject();
+
+            int id = object.getAsJsonPrimitive("id").getAsInt();
+            JsonPrimitive jsonData = object.getAsJsonPrimitive("data");
+            short data = jsonData != null ? jsonData.getAsShort() : 0;
+            String name = object.getAsJsonPrimitive("name").getAsString();
+            JsonPrimitive blockField = object.getAsJsonPrimitive("block");
+            boolean block = blockField != null && blockField.getAsBoolean();
+
+            if (key.indexOf('-') == -1) {
+                int unmappedId;
+                int dataSeparatorIndex = key.indexOf(':');
+                if (dataSeparatorIndex != -1) {
+                    // Include data
+                    short unmappedData = Short.parseShort(key.substring(dataSeparatorIndex + 1));
+                    unmappedId = Integer.parseInt(key.substring(0, dataSeparatorIndex));
+                    unmappedId = (unmappedId << 4) | (unmappedData & 15);
+                } else {
+                    unmappedId = Integer.parseInt(key) << 4;
+                }
+
+                replacementData.put(unmappedId, new MappedLegacyBlockItem(id, data, name, block));
+                return;
             }
-        }
-    }
 
-    private static void addMapping(String key, JsonObject object, Int2ObjectMap<MappedLegacyBlockItem> mappings) {
```

Comment:
Move the code back out to a method, otherwise the constructor looks a bit full/deep

Reply:
Fixed, moved the code below the constructor / removed the static